Source/WebCore/platform/graphics/PathSegmentData.cpp

```diff
@@ -532,43 +532,159 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathRoundedRect& data)
     return ts;
 }
 
-template<class DataType1, class DataType2>
-FloatPoint PathDataComposite<DataType1, DataType2>::calculateEndPoint(const FloatPoint& currentPoint, FloatPoint& lastMoveToPoint) const
+FloatPoint PathDataLine::calculateEndPoint(const FloatPoint&, FloatPoint& lastMoveToPoint) const
 {
-    auto localCurrentPoint = data1.calculateEndPoint(currentPoint, lastMoveToPoint);
-    return data2.calculateEndPoint(localCurrentPoint, lastMoveToPoint);
+    lastMoveToPoint = start;
+    return end;
 }
 
-template<class DataType1, class DataType2>
-void PathDataComposite<DataType1, DataType2>::extendFastBoundingRect(const FloatPoint& currentPoint, const FloatPoint& lastMoveToPoint, FloatRect& boundingRect) const
+void PathDataLine::extendFastBoundingRect(const FloatPoint& currentPoint, const FloatPoint& lastMoveToPoint, FloatRect& boundingRect) const
 {
-    data1.extendFastBoundingRect(currentPoint, lastMoveToPoint, boundingRect);
-    auto localLastMoveToPoint = lastMoveToPoint;
-    auto localCurrentPoint = data1.calculateEndPoint(currentPoint, localLastMoveToPoint);
-    data2.extendFastBoundingRect(localCurrentPoint, lastMoveToPoint, boundingRect);
+    extendBoundingRect(currentPoint, lastMoveToPoint, boundingRect);
+}
+
+void PathDataLine::extendBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    boundingRect.extend(start);
+    boundingRect.extend(end);
+}
+
+void PathDataLine::addToImpl(PathImpl& impl) const
+{
+    impl.moveTo(start);
+    impl.addLineTo(end);
+}
+
+void PathDataLine::applyElements(const PathElementApplier& applier) const
+{
+    applier({ PathElement::Type::MoveToPoint, { start } });
+    applier({ PathElement::Type::AddLineToPoint, { end } });
+}
+
+WTF::TextStream& operator<<(WTF::TextStream& ts, const PathDataLine& data)
+{
+    ts << "move to " << data.start;
+    ts << ", ";
+    ts << "add line to " << data.end;
+    return ts;
+}
+
+FloatPoint PathDataQuadCurve::calculateEndPoint(const FloatPoint&, FloatPoint& lastMoveToPoint) const
+{
+    lastMoveToPoint = start;
+    return endPoint;
+}
+
+void PathDataQuadCurve::extendFastBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    boundingRect.extend(start);
+    boundingRect.extend(controlPoint);
+    boundingRect.extend(endPoint);
+}
+
+void PathDataQuadCurve::extendBoundingRect(const FloatPoint&, const FloatPoint&, FloatRect& boundingRect) const
+{
+    auto extremity = calculateQuadraticExtremity(start, controlPoint, endPoint);
+    boundingRect.extend(start);
+    boundingRect.extend(extremity);
+    boundingRect.extend(endPoint);
```

Comment:
Ditto, re: my other comments — perhaps we can just put this behind a single helper method, or even a new method on `WebCore::FloatRect`? Something like:

```
template<typename... PointType>
void extendBounds(FloatRect& bounds, const PointType&... points)
{
    float minX = std::min({ bounds.x(), points.x()... });
    float minY = std::min({ bounds.y(), points.y()... });
    float maxX = std::max({ bounds.maxX(), points.x()... });
    float maxY = std::max({ bounds.maxY(), points.y()... });
    bounds.setLocation({ minX, minY });
    bounds.setSize({ maxX - minX, maxY - minY });
}
```

...so you can simply do this:

```
extendBounds(boundingRect, currentPoint, controlPoint1, calculateArcToEndPoint(currentPoint, controlPoint1, controlPoint2, radius));
```

or alternately:

```
boundingRect.extend(currentPoint, controlPoint1, calculateArcToEndPoint(currentPoint, controlPoint1, controlPoint2, radius));
```